core/impl/src/main/java/com/blazebit/persistence/impl/AbstractCTECriteriaBuilder.java

```diff
@@ -221,18 +219,38 @@ public CTEInfo createCTEInfo() {
                 if (selectExpression instanceof NullExpression) {
                     // When binding null, we don't have to adapt anything
                 } else if (selectExpression instanceof PathExpression) {
-                    PathExpression pathExpression = (PathExpression) selectExpression;
-                    // Only append the id if it's not already there
-                    if (!idAttribute.getName().equals(pathExpression.getExpressions().get(pathExpression.getExpressions().size() - 1).toString())) {
-                        pathExpression.getExpressions().add(new PropertyExpression(idAttribute.getName()));
+                    boolean firstBinding = true;
+                    PathExpression baseExpression = ((PathExpression) selectExpression).clone(false);
+                    for (String idOrUniqueKeyProp : idOrUniqueKeyProps) {
+                        // For the first id prop, alter the existing expression, for consecutive props
+                        // add additional selects
+                        PathExpression pathExpression = firstBinding ?
+                                ((PathExpression) selectExpression) : baseExpression.clone(false);
+
+                        if (!idOrUniqueKeyProp.equals(pathExpression.getExpressions().get(pathExpression.getExpressions().size() - 1).toString())) {
+                            pathExpression.getExpressions().add(new PropertyExpression(idOrUniqueKeyProp));
+
+                            String nestedAttributeName = attributeName + "_a" + idOrUniqueKeyProp;
```

Comment:
Typo, introduced this to check to ensure that my tests didn't incorrectly pass because I used a "logical" name as alias.